### PR TITLE
Feature/detect defense traversal

### PR DESCRIPTION
**Pull request by JohnGilb**

### CODE REVIEW

File: CompetitionTemplate/src/competition/subsystems/drive/commands/MonitorDefenseTraversalModule.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        
+        // just in case people don't give absolute pitch
+        absolutePitch = Math.abs(absolutePitch);        
+        DefenseState state = DefenseState.NotOnDefense;
+        
+        if (Timer.getFPGATimestamp() - timeOfRecentPitchEvent < defenseTraversalTime.get()) {
+            // we tilted recently!
+            state = DefenseState.RecentlyOnDefense;
+        }
+        else if (absolutePitch > defenseTraversalPitchThreshold.get())
```
Comment (aschokking) [inline CompetitionTemplate/src/competition/subsystems/drive/commands/MonitorDefenseTraversalModule.java]: make this the first check

File: CompetitionTemplate/src/competition/subsystems/drive/PoseSubsystem.java [hunk 2]
```diff
@@ -70,6 +76,9 @@ public void updateCurrentHeading() {
         lastImuHeading = imu.getYaw();
         
         currentHeadingProp.set(currentHeading.getValue());
+
```
Comment (WasabiFan) [inline CompetitionTemplate/src/competition/subsystems/drive/PoseSubsystem.java]: This function should be renamed to reflect the fact that it does more than update heading

File: CompetitionTemplate/src/competition/subsystems/drive/commands/MonitorDefenseTraversalModule.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    double timeOfRecentPitchEvent;
+    DoubleProperty defenseTraversalTime;
+    DoubleProperty defenseTraversalPitchThreshold;
+    StringProperty defenseTraversalState;
+    
+    @Inject
+    public MonitorDefenseTraversalModule(PropertyManager propMan) {
+        timeOfRecentPitchEvent = -99999999;
```
Comment (WasabiFan) [inline CompetitionTemplate/src/competition/subsystems/drive/commands/MonitorDefenseTraversalModule.java]: Uhhhhh.... How about `Double.MIN_VALUE`? Or, if you know that only positive numbers can be returned, -1.

File: CompetitionTemplate/src/competition/subsystems/drive/commands/UpdatePoseCommand.java [hunk 4]
```diff
@@ -1,19 +1,23 @@
-package competition.subsystems.drive.commands;
+    package competition.subsystems.drive.commands;
```
Comment (WasabiFan) [inline CompetitionTemplate/src/competition/subsystems/drive/commands/UpdatePoseCommand.java]: This shouldn't be tabbed in.

File: CompetitionTemplate/src/competition/subsystems/drive/commands/UpdatePoseCommand.java [hunk 5]
```diff
@@ -1,19 +1,23 @@
...
     final PoseSubsystem pose;
+    final MonitorDefenseTraversalModule defenseTraversal;
     
     @Inject
-    public UpdatePoseCommand(PoseSubsystem pose)
+    public UpdatePoseCommand(PoseSubsystem pose, MonitorDefenseTraversalModule defenseTraversal)
     {
+        this.setRunWhenDisabled(true);
         this.pose = pose;
+        this.defenseTraversal = defenseTraversal;
```
Comment (WasabiFan) [inline CompetitionTemplate/src/competition/subsystems/drive/commands/UpdatePoseCommand.java]: The name "defenseTraversal" doesn't really tell me what it is; when we are quickly trying to implement some new functionality, we want it to be clear.

File: CompetitionTemplate/tests/competition/subsystems/drive/MonitorDefenseTraversalTest.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    }
+    
+    @Test
+    public void testSuddenTilt() {
+        assertTrue(module.measureState(50) == DefenseState.OnDefense);
+    }
+    
+    @Test
+    public void testTiltFollowedByFlat() {
+        assertTrue(module.measureState(-50) == DefenseState.OnDefense);
```
Comment (WasabiFan) [inline CompetitionTemplate/tests/competition/subsystems/drive/MonitorDefenseTraversalTest.java]: It's scary that "measyreState" stores state about the value you give it. Can we change it to "updateState" or similar?

**Pull request merged**